### PR TITLE
Groundwork For Updated Scheduler

### PR DESCRIPTION
**Pull request by emaicus**
Provided groundwork for updated scheduler for use in shipping jobs to other machines.

Included the additional info in grading queue files
* Worst case grading time
* Required capabilities
* queue file creation time

Introduced autograding_workers.json int site/config. The file contains information about machines registered to submitty.

### CODE REVIEW

File: bin/regrade.py [hunk 1]
```diff
@@ -88,6 +88,19 @@ def main():
                 my_semester=my_dirs[len(data_dirs)]
                 my_course=my_dirs[len(data_dirs)+1]
                 my_gradeable=my_dirs[len(data_dirs)+3]
+                gradeable_config = os.path.join(data_dir,my_semester,my_course,"config/build/"+"build_"+my_gradeable+".json")
+                with open(gradeable_config, 'r') as build_configuration:
+                    datastore = json.load(build_configuration)
+                    if "required_capabilities" in datastore:
+                        required_capabilities = datastore["required_capabilities"]
+                    else:
+                        required_capabilities = "default"
```
Comment (MasterOdin) [inline bin/regrade.py]: `required_capabilities = datastore.get('required_capabilities', 'default')`

File: bin/regrade.py [hunk 2]
```diff
@@ -88,6 +88,19 @@ def main():
...
+                    if "max_possible_grading_time" in datastore:
+                        max_grading_time = datastore["max_possible_grading_time"]
+                    else:
+                        max_grading_time = -1
```
Comment (MasterOdin) [inline bin/regrade.py]: `max_grading_time = datastore.get('max_possible_grading_time', -1)`

File: site/app/models/Gradeable.php [hunk 3]
```diff
@@ -485,6 +491,21 @@ protected function loadGradeableConfig() {
           $num_textboxes = count($details['textboxes']);
         }
 
+        if (isset($details['required_capabilities'])){
+            $this->required_capabilities = $details['required_capabilities'];
+        }
+        else{
+            $this->required_capabilities = "default";
+        }
```
Comment (MasterOdin) [inline site/app/models/Gradeable.php]: `$this->required_capabilities = $details['required_capabilities'] ?? 'default';`

File: site/app/models/Gradeable.php [hunk 4]
```diff
@@ -485,6 +491,21 @@ protected function loadGradeableConfig() {
...
+
+        if (isset($details['max_possible_grading_time'])){
+            $this->max_possible_grading_time = $details['max_possible_grading_time'];
+        }
+        else{
+           $this->max_possible_grading_time = -1;
+        }
```
Comment (MasterOdin) [inline site/app/models/Gradeable.php]: `$this->max_possible_grading_time = $details['max_possible_grading_time'] ?? -1;`

**Comment by emaicus** [comment]
Making additional edits to change the location of autograding_workers.json. Please wait for additional commit.

**Comment by emaicus** [comment]
Ready to merge.

File: bin/regrade.py [hunk 5]
```diff
@@ -88,6 +88,14 @@ def main():
...
                 my_course=my_dirs[len(data_dirs)+1]
                 my_gradeable=my_dirs[len(data_dirs)+3]
+                gradeable_config = os.path.join(data_dir,my_semester,my_course,"config/build/"+"build_"+my_gradeable+".json")
+                with open(gradeable_config, 'r') as build_configuration:
+                    datastore = json.load(build_configuration)
+                    required_capabilities = datastore.get('required_capabilities', 'default')
+                    max_grading_time = datastore.get('max_possible_grading_time', -1)
+
+                #get the current time
+                queue_time = dateutils.write_submitty_date() + " " + str(dateutils.get_timezone())
```
Comment (bmcutler) [inline bin/regrade.py]: Leave off the get_timezone thing.  All that's needed is in the write_submitty_date.

**Pull request merged**